03-rec-poly/homework-03/hw-03.hs

```diff
@@ -0,0 +1,90 @@
+skips "ABCD"
+skips "hello!"
+skips [1]
+skips [True,False] == [[True,False], [False]]
+skips []           == []
+Note that the output should be the same length as the input.
+-}
+--Exercise 1
+skips :: forall a. [a] -> [[a]]
+skips xs = map (\(idx, _) -> filterIdx idx ps) ps
+  where
+    ps :: [(Int, a)]
+    ps = zip [1 .. length xs] xs
+
+    filterIdx :: Int -> [(Int, a)] -> [a]
+    filterIdx n ls = map snd $ filter (\(idx, _) -> idx `mod` n == 0) ls
+
+
+{-
+Exercise 2 Local maxima
+A local maximum of a list is an element of the list which is strictly
+greater than both the elements immediately before and after it. For example, in the list [2,3,4,1,5], the only local maximum is 4, since it is greater than the elements immediately before and after it (3 and 1). 5 is not a local maximum since there is no element that comes after it.
+Write a function
+localMaxima :: [Integer] -> [Integer]
+which finds all the local maxima in the input list and returns them in order. For example:
+ localMaxima [2,9,5,6,1] == [9,6]
+ localMaxima [2,3,4,1,5] == [4]
+ localMaxima [1,2,3,4,5] == []
+-}
+--Exercise 2
+localMaxima :: [Integer] -> [Integer]
+localMaxima (x : y : z : zs)
+    | y > x && z > y = y : localMaxima (z : zs)
+    | otherwise      = localMaxima (y : z : zs)
+localMaxima _ = []
+
+
+{-
+Exercise 3 Histogram
+For this task, write a function
+histogram :: [Integer] -> String
+which takes as input a list of Integers between 0 and 9 (inclusive), and outputs a vertical histogram showing how many of each number were in the input list. You may assume that the input list does not contain any numbers less than zero or greater than 9 (that is, it does not matter what your function does if the input does contain such numbers). Your output must exactly match the output shown in the examples below.
+cis 194: homework 3 3
+histogram [1,1,1,5] ==
+*
+* **
+ ==========
+ 0123456789
+ histogram [1,4,5,4,6,6,3,4,2,4,9] ==
+*
+* **
+  ******  *
+ ==========
+ 0123456789
+Important note: If you type something like histogram [3,5] at the ghci prompt, you should see something like this:
+ "   * *    \n==========\n0123456789\n"
+This is a textual representation of the String output, including \n escape sequences to indicate newline characters. To actually visualize the histogram as in the examples above, use putStr, for example, putStr (histogram [3,5]).
+-}
+--Exercise 3
+histogram :: [Integer] -> String
+histogram xs = unlines (map (\row -> genLine freqMap row) [mostFreq, mostFreq - 1 .. 1]) ++ "==========\n0123456789\n"
+  where
+    getFreq :: [Integer] -> [Int] 
+    getFreq ls = map (\num -> length $ filter (== num) ls) [0 .. 9]
+
+    freqMap :: [Int]
+    freqMap = getFreq xs
+
+    mostFreq :: Int
+    mostFreq = maximum freqMap
+
+    genLine :: [Int] -> Int -> String
+    genLine freqMap row = map (\freq -> getStr freq row) freqMap
+
+    getStr :: Int -> Int -> Char
+    getStr freq row
+        | freq >= row = '*'
+        | otherwise   = ' '
```

Comment:
1. We should use declarative style, where we start from higher level of abstractions to lower levels.

in the declaration of `histogram`:

```haskell
histogram xs = ... freqMap ... mostFreq, mostFreq - 1 ...
```

Those above variables should be declared immediately after:

```haskell
histogram xs = ... freqMap ... mostFreq, mostFreq - 1 ...
  where
    freqMap = ...
    mostFreq = ...
```

Instead we got:

```haskell
histogram xs = ... freqMap ... mostFreq, mostFreq - 1 ...
  where
    -- Where did this come from??
    getFreq = ...

    -- Oh it comes from here
    freqMap = getFreq ...
    mostFreq = ...
```

This affects readability of the code. A source of confusion could be the ordering. `getFreq` will be in scope even if it is placed after `freqMap` like so:

```haskell
histogram xs = ... freqMap ... mostFreq, mostFreq - 1 ...
  where
    freqMap = getFreq ... -- getFreq still in scope as it is indented at the same level.
    mostFreq = ...

    getFreq = ...
```

However, I would prefer something like this, which makes use of indentation to give us visual indication for definitions of intermediate bindings:

```haskell
histogram xs = ... freqMap ... mostFreq, mostFreq - 1 ...
  where
    freqMap = getFreq ...
      where
        getFreq = ...

    mostFreq = ...
      where
        ...
```

As you can see above, `mostFreq` and `freqMap`, which are the intermediate bindings, are the only two on that indentation level. The other functions are nested deeper under their own respective `where` clauses. This serves the purpose of `scoping` their internal parts.

2. use of lambda here seems like extra noise

```haskell
(\row -> genLine freqMap row)
```

We can just write
```haskell
genLine freqMap
```

3. ` "==========\n0123456789\n"` is a magic number

The comments do provide some sense of what it is. But you should still provide it as a separate binding to declare its purpose:

```haskell
<some-meaningful-name> =  "==========\n0123456789\n"
```

Reply:
@kwannoel Thanks for the feedback, they were detailed and well-demarcated.

I would like to clarify on the level of abstraction that was mentioned at the start of point 1. Are primitives such as `mostFreq :: Int` considered equal in abstraction level vis-a-vis function definitions like `getStr :: Int -> Int -> Char`, since functions are first-class citizens in Haskell. And if they are different, which of the 2 will have a higher level of abstraction (meaning they will be declared first in the `where` clause)?

Thank you.

Reply:
```haskell
eat snack = decreaseHungerBy x
    where
        decreaseHungerBy = f1 f2 f3 ...
            where f1 = ...
                  f2 = ...
                  f3 = ...
        x = b1 b2
            where b1 = ...
                  b2 = ...
```

I consider `decreaseHungerBy` and `x` here to be at the same level of abstraction, regardless of their type signatures.

I guess you can just put them in the order they appear? i.e. `decreaseHungerBy` first then `x` in the `where` clause. But that's if we're being pedantic. Main thing is maintaining them at the same indentation level.

in the above example `f1 f2 f3` and `b1 b2` are implementation details of `decreaseHungerBy` and `x` respectively, that's why I consider them to be more _primitive_ and hence lower levels of abstraction.

Reply:
@kwannoel Thanks alot, this cleared my query up. I think I misunderstood what you were referring to in the term when talking about abstraction levels.